layer/filestore.go

```
@@ -191,6 +201,23 @@ func (fms *fileMetadataStore) GetCacheID(layer ChainID) (string, error) {
 	return content, nil
 }
 
+func (fms *fileMetadataStore) GetForeignSource(layer ChainID) (*distribution.Descriptor, error) {
+	content, err := ioutil.ReadFile(fms.getLayerFilename(layer, "foreign-source"))
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil, nil
+		}
+		return nil, err
+	}
+
+	var ref *distribution.Descriptor
+	err = json.Unmarshal(content, &ref)
+	if err != nil {
+		return nil, err
+	}
+	return ref, err
```

I would probably return a sentinel error value like `ErrNoForeignSource` to return in this case.